### PR TITLE
Make dust-on-leaving-area code less bad

### PR DESCRIPTION
**Pull request by Putnam3145**
## About The Pull Request

See title. The way I did it before was terrible.

## Why It's Good For The Game

Bad code is bad. Good code is good. This is good code.

## Changelog
:cl:
code: Fixed up dusts_on_leaving_area element.
/:cl:

### CODE REVIEW

File: code/datums/elements/dusts_on_leaving_area.dm [hunk 1]
```diff
@@ -1,28 +1,21 @@
...
 	var/list/area_types = list()
 
 /datum/element/dusts_on_leaving_area/Attach(datum/target,types)
 	. = ..()
 	if(!ismob(target))
 		return ELEMENT_INCOMPATIBLE
-	attached_mobs += target
 	area_types = types
-	START_PROCESSING(SSprocessing,src)
+	RegisterSignal(target,COMSIG_AREA_ENTERED,.proc/check_dust)
```
Comment (Ghommie) [inline code/datums/elements/dusts_on_leaving_area.dm]:
```suggestion
	RegisterSignal(target, COMSIG_ENTER_AREA, .proc/check_dust)
```

File: code/datums/elements/dusts_on_leaving_area.dm [hunk 2]
```diff
@@ -1,28 +1,21 @@
...
 
 /datum/element/dusts_on_leaving_area/Detach(mob/M)
 	. = ..()
-	if(M in attached_mobs)
-		attached_mobs -= M
-	if(!attached_mobs.len)
-		STOP_PROCESSING(SSprocessing,src)
+	UnregisterSignal(M,COMSIG_AREA_ENTERED)
```
Comment (Ghommie) [inline code/datums/elements/dusts_on_leaving_area.dm]:
```suggestion
	UnregisterSignal(M, COMSIG_ENTER_AREA)
```

File: code/datums/elements/dusts_on_leaving_area.dm [hunk 3]
```diff
@@ -1,28 +1,21 @@
...
-		var/mob/M = m
-		var/area/A = get_area(M)
-		if(!(A.type in area_types))
-			M.dust(force = TRUE)
-			Detach(M)
+/datum/element/dusts_on_leaving_area/proc/check_dust(mob/M)
+	var/area/A = get_area(M)
+	if(!(A.type in area_types))
+		M.dust(force = TRUE)
+		Detach(M)
```
Comment (Ghommie) [inline code/datums/elements/dusts_on_leaving_area.dm]: redundant, `ELEMENT_DETACH` makes so the element is detached from deleted datums already.

**Review by Ghommie — CHANGES_REQUESTED** [review]

File: code/datums/elements/dusts_on_leaving_area.dm (hunk 2 above)
Comment (Putnam3145) [reply code/datums/elements/dusts_on_leaving_area.dm]: this is correct but unfortunately COMSIG_ENTER_AREA also has the *area* as the argument, which makes it a mite harder to work with.

Comment (Putnam3145) [reply code/datums/elements/dusts_on_leaving_area.dm]: actually no i have a solution for this

File: code/game/area/areas.dm [hunk 4]
```diff
@@ -493,8 +493,8 @@ GLOBAL_LIST_EMPTY(teleportlocs)
 
 /area/Entered(atom/movable/M)
 	set waitfor = FALSE
-	SEND_SIGNAL(src, COMSIG_AREA_ENTERED, M)
-	SEND_SIGNAL(M, COMSIG_ENTER_AREA, src) //The atom that enters the area
+	SEND_SIGNAL(src, COMSIG_AREA_ENTERED, M, src)
+	SEND_SIGNAL(M, COMSIG_ENTER_AREA, src, M) //The atom that enters the area
```
Comment (Ghommie) [inline code/game/area/areas.dm]: Dude, the first argument of the macro is also the first in the signal procs args, this is redundant.

File: code/game/area/areas.dm [hunk 5]
```diff
@@ -519,8 +519,8 @@ GLOBAL_LIST_EMPTY(teleportlocs)
 			addtimer(CALLBACK(L.client, /client/proc/ResetAmbiencePlayed), 600)
 
 /area/Exited(atom/movable/M)
-	SEND_SIGNAL(src, COMSIG_AREA_EXITED, M)
-	SEND_SIGNAL(M, COMSIG_EXIT_AREA, src) //The atom that exits the area
+	SEND_SIGNAL(src, COMSIG_AREA_EXITED, M, src)
+	SEND_SIGNAL(M, COMSIG_EXIT_AREA, src, M) //The atom that exits the area
```
Comment (Ghommie) [inline code/game/area/areas.dm]: Same.

File: code/datums/elements/dusts_on_leaving_area.dm (hunk 2 above)
Comment (Ghommie) [reply code/datums/elements/dusts_on_leaving_area.dm]: You are reading it wrong, COMSIG_ENTER_AREA is the signal sent to the mob, COMSIG_AREA_ENTERED is sent to the area.

**Review by Ghommie — CHANGES_REQUESTED** [review]

**Review by Ghommie — APPROVED** [review]

**Pull request merged**